### ISSUE TITLE
Redundant cast inside expression changes result in 0.20.23

### ISSUE THREAD

**Issue by lukeshingles** [pola-rs/polars#15952]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

df = pl.DataFrame({"x": range(5)})
print(df.select((pl.col("x") * 1.0e9 / 10)))
print(df.select((pl.col("x") * 1.0e9 / 10).cast(pl.Float64)))
```

### Log output

```shell
shape: (5, 1)
┌─────┐
│ x   │
│ --- │
│ f64 │
╞═════╡
│ 0.0 │
│ 1e8 │
│ 2e8 │
│ 3e8 │
│ 4e8 │
└─────┘
shape: (5, 1)
┌──────┐
│ x    │
│ ---  │
│ f64  │
╞══════╡
│ 0.0  │
│ 1e8  │
│ 2e8  │
│ null │
│ null │
```

### Issue description

This seems to be some kind of calculation range issue that gets triggered by the integer in the expression (replacing 10 with 10.0 prevents the issue) and the cast to float as part of the expression, which should not change the result, since the column resulting from the expression is already Float64.

### Expected behavior

In 0.20.22, the two DataFrames as equal, as expected.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.23
Index type:           UInt32
Platform:             macOS-14.4.1-arm64-arm-64bit
Python:               3.12.3 (main, Apr  9 2024, 08:09:14) [Clang 15.0.0 (clang-1500.3.9.4)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               24.2.1
hvplot:               <not installed>
matplotlib:           3.8.4
nest_asyncio:         1.6.0
numpy:                1.26.4
openpyxl:             <not installed>
pandas:               2.2.2
pyarrow:              16.0.0
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

**Comment by lukeshingles** [comment]
Sorry if this is a duplicate. This  seems to be related to #15951, which I've confirmed also became an issue with polars 0.20.23.

**Comment by lukeshingles** [comment]
This is fixed in polars 0.20.24.

**Comment by TobiasDummschat** [comment]
I could reproduce this with version 0.20.23, but it seems to be fixed in version 0.20.24 which released today.
